src/main/java/uk/gov/hmcts/juror/api/moj/repository/letter/court/DeferralDeniedLetterListRepositoryImpl.java

```diff
@@ -1,83 +1,194 @@
+        JPAQuery<Tuple> jpaQuery = buildBaseQuery(owner);
 
         filterEligibleLetterSearchCriteria(jpaQuery, searchCriteria);
 
         orderQueryResults(jpaQuery, searchCriteria.includePrinted());
 
-        return jpaQuery.fetch();
+        List<Tuple> results = jpaQuery.fetch();
+
+        List<DeferralDeniedLetterList> deferralDeniedLetterLists = results.stream()
+            .map(tuple -> DeferralDeniedLetterList.builder()
+                .poolNumber(tuple.get(QJurorPool.jurorPool.pool.poolNumber))
+                .jurorNumber(tuple.get(QJuror.juror.jurorNumber))
+                .firstName(tuple.get(QJuror.juror.firstName))
+                .lastName(tuple.get(QJuror.juror.lastName))
+                .postcode(tuple.get(QJuror.juror.postcode))
+                .status(tuple.get(QJurorStatus.jurorStatus.statusDesc))
+                .otherInformation(tuple.get(QJurorHistory.jurorHistory.otherInformation))
+                .refusalDate(tuple.get(QJurorHistory.jurorHistory.dateCreatedDateOnly))
+                .isActive(tuple.get(QJurorPool.jurorPool.isActive))
+                .build())
+            .collect(Collectors.toList());
+
+
+        // for each juror select the most recent deferred date and remove the other entries
+        // this is to ensure that only the most recent deferral date is returned
+        ConcurrentHashMap<String, DeferralDeniedLetterList> deferralDeniedLetterListMap = new ConcurrentHashMap<>();
+        deferralDeniedLetterLists.forEach(deferralDeniedLetterList -> {
+            String jurorNumber = deferralDeniedLetterList.getJurorNumber();
+            if (deferralDeniedLetterListMap.containsKey(jurorNumber)) {
+                DeferralDeniedLetterList existingDeferralDeniedLetterList =
+                    deferralDeniedLetterListMap.get(jurorNumber);
+                if (deferralDeniedLetterList.getRefusalDate()
+                    .isAfter(existingDeferralDeniedLetterList.getRefusalDate())) {
+                    deferralDeniedLetterListMap.put(jurorNumber, deferralDeniedLetterList);
+                }
+            } else {
+                deferralDeniedLetterListMap.put(jurorNumber, deferralDeniedLetterList);
+            }
+        });
+
+        List<String> poolNumbers =
+            deferralDeniedLetterLists.stream().map(DeferralDeniedLetterList::getPoolNumber).toList();
+
+        // run query to get the most recent deferral denied date for each juror
+        List<Tuple> printedDates = getPrintedDate(owner, poolNumbers);
+
+        // update the deferral denied printed date for each juror
+        printedDates.forEach(tuple -> {
+            String jurorNumber = tuple.get(QJuror.juror.jurorNumber);
+            if (deferralDeniedLetterListMap.containsKey(jurorNumber)) {
+                DeferralDeniedLetterList deferralDeniedLetterList = deferralDeniedLetterListMap.get(jurorNumber);
+                LocalDateTime datePrinted = tuple.get(QJurorHistory.jurorHistory.dateCreated);
```

Comment:
Can you grab convert toLocalDate() here so it is done once

Reply:
done